libbeat/publisher/queue/memqueue/eventloop.go

```diff
@@ -323,11 +321,28 @@ func (l *bufferingEventLoop) run() {
 }
 
 func (l *bufferingEventLoop) handleMetricsRequest(req *metricsRequest) {
-	req.responseChan <- memQueueMetrics{currentQueueSize: l.eventCount, occupiedRead: l.unackedEventCount}
+	oldestEntryID := l.nextEntryID
+	// If we have ACKs pending, the oldest id is the first one.
+	// Otherwise, it's still in the flush list. If both are empty,
+	// then we report the "oldest" as the next id that will be assigned.
+	if !l.pendingACKs.empty() {
+		oldestEntryID = l.pendingACKs.head.entries[0].id
+	} else if !l.flushList.empty() {
+		oldestEntryID = l.flushList.head.entries[0].id
+	}
+	req.responseChan <- memQueueMetrics{
+		currentQueueSize: l.eventCount,
+		occupiedRead:     l.unackedEventCount,
+		oldestEntryID:    oldestEntryID,
+	}
 }
 
 func (l *bufferingEventLoop) handleInsert(req *pushRequest) {
-	if l.insert(req) {
+	if l.insert(req, l.nextEntryID) {
+		// Send back the new event id.
+		req.resp <- l.nextEntryID
```

Comment:
With these response channels, is there any chance that the receiving side has been cancelled or is otherwise unresponsive? Do we need to handle cases where this send blocks unexpectedly?

Reply:
No (though I could add more comments):
- these channels are only accessible internally to the queue
- any channel-based API call immediately listens on the response channel once the request is sent
- the response channels have buffer size 1 so the response can't block the main queue goroutine.
- once the queue accepts a request it will always send a response